crates/bevy_ecs/src/event.rs

```diff
@@ -119,50 +119,104 @@ enum State {
 ///
 /// [`AppBuilder::add_event`]: https://docs.rs/bevy/*/bevy/app/struct.AppBuilder.html#method.add_event
 #[derive(Debug)]
-pub struct Events<T> {
-    events_a: Vec<EventInstance<T>>,
-    events_b: Vec<EventInstance<T>>,
+pub struct Events<T: Event> {
+    events_a: T::Storage,
+    events_b: T::Storage,
     a_start_event_count: usize,
     b_start_event_count: usize,
     event_count: usize,
     state: State,
 }
 
-impl<T> Default for Events<T> {
+pub trait Event: Sized + Component {
+    type Storage: for<'a> Storage<'a, Item = EventInstance<Self>>
+        + Component
+        + std::ops::DerefMut<Target = [EventInstance<Self>]>
+        + Default;
+}
+
+pub trait Storage<'a> {
+    type Item: 'a;
+    type DrainIter: DoubleEndedIterator<Item = Self::Item> + 'a;
```

Comment:
Is it worth loosening this requirement from just `DoubleEndedIterator`s?
If we're now allowing custom storage types for `Events`, we should maybe support more than just these types.
For example, if I don't want any duplicate events, I may want to make a `Storage` type which uses a `HashSet` internally, but `hash_set::Iter` does not implement `DoubleEndedIter`.

Reply:
This trait is a mostly internal trait, which isn't meant to be implemented. HashSet would also fail because it doesn't match the `Deref<[_]>` criteria.